File: src/main/kotlin/me/alvr/asfui/views/Bots.kt
package me.alvr.asfui.views

import javafx.scene.Parent
import javafx.scene.control.Button
import javafx.scene.control.ComboBox
import me.alvr.asfui.ASFProcess
import me.alvr.asfui.Command
import tornadofx.Fragment
import tornadofx.action
import tornadofx.enableWhen
import tornadofx.selectedItem

class Bots : Fragment() {
    override val root: Parent by fxml("/bots.fxml")

    private val botsStart: Button by fxid("bots_start")
    private val botsStartAll: Button by fxid("bots_startall")
    private val botsStop: Button by fxid("bots_stop")
    private val botsPause: Button by fxid("bots_pause")
    private val botsResume: Button by fxid("bots_resume")
    private val botsPassword: Button by fxid("bots_password")
    private val botsStatus: Button by fxid("bots_status")
    private val botsStatusAll: Button by fxid("bots_statusall")

    private val bots: ComboBox<String> by param()
    private var bot = bots.selectedItem!!

    init {
        bots.valueProperty().addListener { _, _, newBot ->
            bot = newBot
        }

        botsStart.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.START, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsStartAll.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.START_ALL)
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsStop.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.STOP, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsPause.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.PAUSE, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsResume.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.RESUME, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsPassword.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.PASSWORD, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsStatus.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.generateCommand(Command.STATUS, bot))
                }
            }
            enableWhen(ASFProcess.started)
        }

        botsStatusAll.apply {
            action {
                runAsync {
                    Command.sendCommand(Command.STATUS_ALL)
                }
            }
            enableWhen(ASFProcess.started)
        }
    }
}

File: src/views/bots.py
import threading

from PyQt5 import uic
from PyQt5.QtWidgets import QWidget

from utils.command import generate_command, START_ALL, send_command, START, STATUS_ALL, STATUS, PASSWORD, RESUME, \
    PAUSE, STOP
from utils.resources import resource_path


class Bots(QWidget):
    def __init__(self, bots):
        super().__init__()

        uic.loadUi(resource_path('src/resources/ui/bots.ui'), self)

        self.bots = bots
        self.bots_start.clicked.connect(self.start)
        self.bots_start_all.clicked.connect(self.start_all)
        self.bots_stop.clicked.connect(self.stop)
        self.bots_pause.clicked.connect(self.pause)
        self.bots_resume.clicked.connect(self.resume)
        self.bots_password.clicked.connect(self.password)
        self.bots_status.clicked.connect(self.status)
        self.bots_status_all.clicked.connect(self.status_all)

    def start(self):
        command = generate_command(START, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def start_all(self):
        command = generate_command(START_ALL)
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def stop(self):
        command = generate_command(STOP, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def pause(self):
        command = generate_command(PAUSE, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def resume(self):
        command = generate_command(RESUME, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def password(self):
        command = generate_command(PASSWORD, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def status(self):
        command = generate_command(STATUS, self.bots.currentText())
        threading.Thread(target=send_command, args=(command,), daemon=True).start()

    def status_all(self):
        command = generate_command(STATUS_ALL)
        threading.Thread(target=send_command, args=(command,), daemon=True).start()